Fix GetColumnCount for wxLC_LIST style

Kvaz1r opened the pull request:
Fix: #22482 

Unlike wxMSW the generic version of GetColumnCount returns 0 for wxLC_LIST style that prevent correct handling SendNotify at least under wxUniv. 

I can't test other ports that use generic version so don't know does the bug exist there. If not and such changes may break something I'll update GetColumnCount in PR only for wxUniv with this:

```
int wxGenericListCtrl::GetColumnCount() const
{
#ifdef __WXUNIVERSAL__
    if ( HasFlag(wxLC_LIST) )
        return 1;
#endif
    return m_mainWin->GetColumnCount();
}
```

Kvaz1r committed:
Fix GetColumnCount for wxLC_LIST style

Kvaz1r committed:
Add tests for wxListCtrl ColumnCount

Kvaz1r commented:
> Thanks, I think we should change this for all ports, but I'd like to add a new test to `tests/controls/listctrltest.cpp` verifying that they all behave consistently here. Please update it if you can, otherwise I'll try to do it myself later. TIA!


Updated

Kvaz1r commented:
> Thanks, I think we should change this for all ports, but I'd like to add a new test to `tests/controls/listctrltest.cpp` verifying that they all behave consistently here. Please update it if you can, otherwise I'll try to do it myself later. TIA!


Updated

vadz committed:
Micro optimization in wxListMainWindow::SendNotify()

Just call GetListCtrl() once instead of calling both it and GetParent()
multiple times.

No real changes.

vadz committed:
Simplify the code in just added ListCtrlTestCase::ColumnCount()

Don't call tearDown() explicitly, we'll get rid of this function sooner
or later, hopefully, so don't add extra calls to it and just delete the
window directly, nothing else should be needed here.

No real changes.

vadz committed:
Extend wxListCtrl::GetColumnCount() tests to all modes

Check that it consistently returns 0 in the wxLC_ICON and
wxLC_SMALL_ICON modes under all platforms.

Document the behaviour of this function in all modes.

vadz commented:
Thanks!

I've added a couple more tests for the other modes too. I'll merge it once the CI builds pass.

vadz commented:
Hmm, it looks like calling `GetListCtrl()` from `SendNotify()` is a bad idea because this fails when it's called when deleting all items from the dtor, so the apparently innocuous [3eee299](https://github.com/wxWidgets/wxWidgets/pull/22497/commits/3eee299cd317182fc0e460d4fcb1dd1e8e7c5558) broke everything, sorry. Will fix it soon.

vadz committed:
Fix assert failure in SendNotify()

Destroy wxGenericListCtrl children earlier, so that this happens while
this object is still a wxListCtrl and not just a wxWindow, to fix an
assert triggered by the changes of 3eee299cd3 (Micro optimization in
wxListMainWindow::SendNotify(), 2022-06-08).
